### PR TITLE
[gen_l10n] Improvements to `gen_l10n`

### LINKED ISSUE #115501: i10n_gen errors/warnings improvement suggestions + need testing
~~After the change here: https://github.com/flutter/flutter/pull/112390, building the new Flutter gallery for iOS fails at tip-of-tree~~. [eta: the build failure was a different issue, filed a separate issue.]

I see some issues with https://github.com/flutter/flutter/pull/112390.
1. Lots of commented out code. If a fast-follow had been landed as @HansMuller mentioned, then that would have been okay, but it's been 11 days or so.
2. I'm not seeing a test for this particular error path: `The plural part...`.
3. The tests that do exist, don't seem to be testing that the second two lines of the error message print something useful.
4. The second part of the message is coming through empty for me, and the third part, which is supposed to have the file and line number are empty, so I can't track down the issue with flutter/gallery.
5. nit: As mentioned on the PR, "overrided" should be "overridden".

cc @thkim1011 @HansMuller @christopherfujino 

There's a separate issue, which seems to be that the new Xcode error handling code seems to treat `printWarning`s from i10n_gen as fatal, both with and without https://github.com/flutter/flutter/pull/112390. cc @jmagman

### PR DESCRIPTION
**Pull request by thkim1011**
This PR contains several changes to the `gen_l10n` tool. 

## Optional placeholder metadata ✅
* Currently, all placeholders are required to be declared in a separate "message metadata" JSON object as a sibling of the "message" JSON object. For example,
```json
{
  "helloWorld": "Hello {name}!",
  "@helloWorld": {
    "placeholder: {
      "name": {}
    }
  }
}
```
* This PR implements a way to make it optional to explicitly declare the placeholders by appending each undeclared placeholder (in every language) to the end of the placeholders list in alphabetical order.

### Tests
Ran this on the flutter-wonderous-app which successfully generated the following code.
```dart
/// The translations for English (`en`).
class AppLocalizationsEn extends AppLocalizations {
  AppLocalizationsEn([String locale = 'en']) : super(locale);
...
  @override
  String collectionLabelCount(Object count, Object total) {
    return '$count of $total';
  }
...
```
Also, see the following tests.
* `should automatically infer placeholders that are not explicitly defined`.
* `should automatically infer plural placeholders that are not explicitly defined`.
* `should automatically infer select placeholders that are not explicitly defined`.

## Lexer fixes ✅
* Currently, the lexer treats "plural" and "select" as a keyword.
* This PR changes this so that if used right after an open brace and we are currently parsing an expression, then treat it as a identifier with value "plural" and "select".

### Tests
Ran on flutter-wonderous-app. The following generates a function which uses the identifier "plural" as a placeholder.
```dart
  @override
  String newlyDiscoveredSemanticNew(Object count, Object plural) {
    return '$count new item$plural to explore. Scroll to new item';
  }
```
Also, see the following test.
* `lexer identifier names can be "select" or "plural"`

## More consistent warnings and errors ✅
* There are some issues with errors where only one error is shown at a time when there may be multiple **syntax** errors across multiple messages and languages.
* Furthermore, the warning for when plural parts are overridden does not follow the same format as the errors.
* This PR addresses both of these issues.
### Tests
Ran this on flutter/gallery and turned on `use-escaping` to generate syntax errors. It now lists the syntax error for each message as opposed to breaking after encountering a single error.
```
[intl_fr.arb:demoTwoPaneItemDetails] ICU Lexing Error: Unmatched single quotes.
    Détails de l'élément {value}
                ^
[intl_fr_CA.arb:demoTwoPaneItemDetails] ICU Lexing Error: Unmatched single quotes.
    Détails de l'élément {value}
                ^
[intl_fr_CH.arb:demoTwoPaneItemDetails] ICU Lexing Error: Unmatched single quotes.
    Détails de l'élément {value}
                ^
Found syntax errors.
```
Also, see the following tests.
* `warnings are generated when plural parts are repeated`.
* `multiple messages with syntax error all log their errors`.
* 
## Change generated "helper functions" to temporary variables. ✅
* Just realized that the helper functions that get generated do not actually have to be functions, they're all used only once. Switching to variables instead also should avoid some bugs that may pop up when passing arguments into helper function parameters incorrectly. 
### Tests
`gen_l10n.dart` which is the full E2E test that spawns the `flutter gen-l10n` process, passes.
Also, generated the code for the message `  "selectInPlural": "{count, plural, =1{{gender, select, male{he} female{she} other{they}}} other{they}}",` with the generated code:
```
  @override
  String selectInPlural(String gender, num count) {
    String _temp0 = intl.Intl.selectLogic(
      gender,
      {
        'male': 'he',
        'female': 'she',
        'other': 'they',
      },
    );
    String _temp1 = intl.Intl.pluralLogic(
      count,
      locale: localeName,
      other: 'they',
      one: '$_temp0',
    );
    return '$_temp1';
  }
```

## Add option to suppress warnings ✅
* Finally, the last change is an option to suppress warnings.

### Tests
Tested this on flutter/gallery which currently generates a lot of warning due to the generated `*.arb` files (working on a fix internally). Adding `suppress-warnings: true` to `l10n.yaml` successfully suppresses the warnings.

Also, see the following test.
* `suppress warnings flag actually suppresses warnings`.

Fixes #115501.

### CODE REVIEW

File: packages/flutter_tools/lib/src/localizations/gen_l10n.dart [hunk 1]
```diff
@@ -1153,17 +1153,7 @@ class LocalizationsGenerator {
...
-          if (placeholder == null) {
-            throw L10nParserException(
-              'Make sure that the specified placeholder is defined in your arb file.',
-              _inputFileNames[locale]!,
-              message.resourceId,
-              translationForMessage,
-              identifier.positionInMessage,
-            );
-          }
+          final Placeholder placeholder = message.placeholders[identifier.value]!;
```
Comment (Piinks) [inline packages/flutter_tools/lib/src/localizations/gen_l10n.dart:1156]: Do we know that the `!` will always find a value that is not null here and elsewhere in this file where it is being added?

File: packages/flutter_tools/lib/src/localizations/gen_l10n_types.dart [hunk 2]
```diff
@@ -149,6 +150,19 @@ ${List<String>.filled(4 + fileName.length + messageId.length + charNumber, ' ').
   final int charNumber;
 }
```
Comment (Piinks) [inline packages/flutter_tools/lib/src/localizations/gen_l10n_types.dart:152]: Can you add a comment or two about what this is and where it is used?

File: packages/flutter_tools/test/general.shard/generate_localizations_test.dart [hunk 3]
```diff
@@ -1598,30 +1598,23 @@ import 'output-localization-file_en.dart' deferred as output-localization-file_e
...
+          templateArbFileName: defaultTemplateArbFileName,
+          outputFileString: defaultOutputFileString,
+          classNameString: defaultClassNameString,
+          logger: logger,
+        )
+          ..loadResources()
+          ..writeOutputFiles();
+        expect(logger.hadWarningOutput, isTrue);
+        expect(logger.warningText, contains('''
+[app_en.arb:helloWorld] ICU Syntax Warning: Missing placeholder declaration. Defaulting to using the message as is without parsing.
```
Comment (Piinks) [inline packages/flutter_tools/test/general.shard/generate_localizations_test.dart:1614]: Can you check the full error messages in these tests? The issue mentions parts of the error message missing. You could probably define a template for the message and re-use it to simplify the test code here.

Comment (Piinks) [reply packages/flutter_tools/test/general.shard/generate_localizations_test.dart:1614]: From the issue

> The tests that do exist, don't seem to be testing that the second two lines of the error message print something useful.
The second part of the message is coming through empty for me, and the third part, which is supposed to have the file and line number are empty, so I can't track down the issue with flutter/gallery.

**Review by Piinks — COMMENTED** [review]
Following up on comments from https://github.com/flutter/flutter/issues/115501

> I'm not seeing a test for this particular error path: The plural part....

Is that added here? Also there at gen_l10n.dart line#1204, there is still the overridden typo.

The issue also mentions a line number, is that supported here?

File: packages/flutter_tools/lib/src/localizations/gen_l10n.dart (hunk 1 above)
Comment (thkim1011) [reply packages/flutter_tools/lib/src/localizations/gen_l10n.dart]: Yes! In the `Message` class where all the parsing and type inference for placeholders happen, we make sure that all types are correct and that all placeholders used in messages exist in the placeholders map.

File: packages/flutter_tools/test/general.shard/generate_localizations_test.dart (hunk 3 above)
Comment (thkim1011) [reply packages/flutter_tools/test/general.shard/generate_localizations_test.dart]: I wasn't able to reproduce that behavior, but the test is now checking for the full error message. I also ran `flutter run` in `flutter/gallery` and was able to see the warnings.

File: packages/flutter_tools/lib/src/localizations/gen_l10n_types.dart (hunk 2 above)
Comment (thkim1011) [reply packages/flutter_tools/lib/src/localizations/gen_l10n_types.dart:153]: Fixed!

**Review by HansMuller — APPROVED** [review]
This set of improvements should help with both backwards compatibility and ease of use. NICE.

File: packages/flutter_tools/lib/src/localizations/gen_l10n.dart [hunk 4]
```diff
@@ -1196,26 +1162,22 @@ class LocalizationsGenerator {
...
-            } else {
+              final String pluralPartExpression = generateVariables(pluralMessage);
+              pluralLogicArgs[pluralCases[pluralCase]!] = '      ${pluralCases[pluralCase]}: $pluralPartExpression,';
+            } else if (!suppressWarnings) {
               logger.printWarning('''
-The plural part specified below is overrided by a later plural part.
-$translationForMessage
-${Parser.indentForError(pluralPart.positionInMessage)}
-''');
+[${_inputFileNames[locale]}:${message.resourceId}] ICU Syntax Warning: The plural part specified below is overriden by a later plural part.
```
Comment (Piinks) [inline packages/flutter_tools/lib/src/localizations/gen_l10n.dart:1169]: Nit: There are 2 d's

`overridden`

**Review by Piinks — APPROVED** [review]
LGTM with a typo to fix

File: packages/flutter_tools/lib/src/localizations/gen_l10n.dart (hunk 4 above)
Comment (thkim1011) [reply packages/flutter_tools/lib/src/localizations/gen_l10n.dart]: Fixed!

**Pull request merged**